### PR TITLE
[#4] Implement errors function

### LINKED ISSUE #4: Implement errors function
- [x] Move `SourcePosition` newtype
- [x] Move `toSourcePosition` function
- [x] Introduce `AppError`:
    ```haskell
    data AppError e = AppError  -- TODO: better name? `WithCallStack`? ErrorWithSource?
        { appErrorCallStack :: !SourcePosition
        , appErrorType      :: !e
        }
- [x] Write `AppException`
- [x] Write `WithError`
- [x] Write `throwError` for `AppError`

### PR DESCRIPTION
**Pull request by chshersh**
Resolves #4

### CODE REVIEW

File: src/CakeSlayer/Error.hs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    } deriving stock (Show, Eq, Functor)
+
+{- | Specialized version of 'E.throwError' that attaches source code position of
+the place where this error was thrown.
+-}
+throwError :: WithError err m => err -> m a
+throwError = E.throwError . ErrorWithSource (toSourcePosition callStack)
+{-# INLINE throwError #-}
+
+newtype SourcePosition = SourcePosition Text
```
Comment (vrom911) [inline src/CakeSlayer/Error.hs]: We usually using named fields for newtypes, I guess we should apply this rule here too

File: src/CakeSlayer/Error.hs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+       , toSourcePosition
+       ) where
+
+import Control.Monad.Except (MonadError)
+import GHC.Stack (SrcLoc (SrcLoc, srcLocModule, srcLocStartLine))
+
+import qualified Control.Monad.Except as E (throwError)
+
+
+{- | Type alias for errors that have acess to 'CallStack'. Specicialise this
```
Comment (vrom911) [inline src/CakeSlayer/Error.hs]:
```suggestion
{- | Type alias for errors that has access to 'CallStack'. Specialise this
```

File: src/CakeSlayer/Error.hs (hunk 1 above)
Comment (vrom911) [reply src/CakeSlayer/Error.hs]: And also let's add haddock comment as we exporting this.

File: src/CakeSlayer/Error.hs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+{- | Exception wrapper around 'ErrorWithSource'. Useful when you need to throw/catch
+'ErrorWithSource' as 'Exception'.
+-}
+newtype AppException err = AppException
+    { unAppException :: ErrorWithSource err
+    } deriving stock (Show)
+      deriving anyclass (Exception)
+
+{- | Helper to convert 'err' into something that can be thrown
```
Comment (vrom911) [inline src/CakeSlayer/Error.hs]:
```suggestion
{- | Helper to convert @err@ into something that can be thrown
```

**Comment by chshersh** [comment]
@vrom911 Good idea to add example of the data type! Added. And also improved code a bit more.

**Pull request merged**